crates/sc-consensus-subspace/src/archiver.rs

```diff
@@ -783,20 +783,39 @@ where
     AS: AuxStore + Send + Sync + 'static,
     SO: SyncOracle + Send + Sync + 'static,
 {
-    let InitializedArchiver {
-        confirmation_depth_k,
-        mut archiver,
-        older_archived_segments,
-        best_archived_block: (mut best_archived_block_hash, mut best_archived_block_number),
-    } = initialize_archiver(&segment_headers_store, subspace_link, client.as_ref())?;
-
-    let mut block_importing_notification_stream = subspace_link
-        .block_importing_notification_stream
-        .subscribe();
-    let archived_segment_notification_sender =
-        subspace_link.archived_segment_notification_sender.clone();
+    let maybe_archiver = if segment_headers_store.max_segment_index().is_none() {
+        Some(initialize_archiver(
+            &segment_headers_store,
+            subspace_link.clone(),
+            client.as_ref(),
+        )?)
+    } else {
+        None
+    };
 
     Ok(async move {
+        let archiver = match maybe_archiver {
+            Some(archiver) => archiver,
+            None => initialize_archiver(
+                &segment_headers_store,
+                subspace_link.clone(),
+                client.as_ref(),
+            )?,
+        };
+
+        let InitializedArchiver {
+            confirmation_depth_k,
+            mut archiver,
+            older_archived_segments,
+            best_archived_block: (mut best_archived_block_hash, mut best_archived_block_number),
+        } = archiver;
+
+        let mut block_importing_notification_stream = subspace_link
+            .block_importing_notification_stream
+            .subscribe();
```

Comment:
I have intentionally shown both in pseudo-code and in out call that this subscription must happen outside of this future, otherwise we will have a race condition where block will be imported, but this subscription will not capture it because it is created afterwards.